### PR TITLE
Automatically logout sessions when password change

### LINKED ISSUE #6384: Should not keep user logged in, if password was changed on his account.
# Ghost should not keep user logged in, if password was changed on his account.
### Issue Summary

When one human changes password for specific user, any other people, who are logged in that user aren't logged out but still logged in and can still do anything.
### Steps to Reproduce (2 humans)
1. One person logs in your website with your account, which is powered by Ghost,
2. You change your password,
3. Another person who has logged in is still there and can do any kind of stuff.

Because of that, I think, ghost needs to log out all people, who are logged in one account, where password was changed, except of the one who changed password.
### Technical details:
- Ghost Version: 0.7.5
- Client OS: Windows 8.1
- Server OS: Ubuntu 15.10, with nginx that adds SSL and some custom error pages, inside Microsoft Azure
- Node Version: 0.10.40
- Browser: Google Chrome Version 48.0.2564.82 m
- Database: SQLite

### PR DESCRIPTION
**Pull request by ramiel**
closes #6384
- If a user change is own password, all the session belonging to him are logged out except the current
- If an owner change the password of another user, all the session belonging to this user are logged out
- Refactored some tests for the new behaviour

### CODE REVIEW

**Comment by acburdine** [comment]
Hey @ramiel! At first glance this looks like a better approach to me :+1: However, your commits seem to have merge conflicts. Would you mind updating them? :smile: 

Also, would it be possible to have some sort of an option to specify more tokens generated than just one, instead of just automatically doing it for every test? Not sure of exactly how much of an impact this will have on testing speed, but just the same I feel like creating multiple tokens on every test is needlessly complicated. I could be completely wrong, and as such will defer to @sebgie's judgement.

**Comment by ramiel** [comment]
The tests fails when the DB is postgres and node is not on version 0.12...which is odd!I'm investigating on it :)
For the token creations: I create two token for each situation where a test is need (when somebody change his own password or when an owner change a password for someone else).
What I can do is to create the tokens just for this two test. If I did not this, is just because I noticed you never create data ad-hoc for a single test (at least in the test suite I modified). Do you mean this with your comment?

**Comment by acburdine** [comment]
That's kind of what I meant, and I'm not sure if it's possible to do without a lot of unnecessary complexity. I just wasn't sure what kind of an impact it would have on testing speed. However, something you could do is just combine all those inserts into a single query, e.g.

```javascript
knex('accesstokens').insert([
    DataGenerator.forKnex.createToken({id: 4, user_id: 3}),
    DataGenerator.forKnex.createToken({id: 5, user_id: 3}),
    // etc...
]);
```

Again, not sure what the differences would be in speed, but this would at least make it cleaner.

**Comment by ramiel** [comment]
Well, probably is the case to insert the tokens only for the tests which need them and not in the setup method of all the suite. What do you think?

**Comment by acburdine** [comment]
If that's possible then go for it :smile:

**Comment by ramiel** [comment]
Absolutely!Tomorrow :D

File: core/test/utils/index.js [hunk 1]
```diff
@@ -251,12 +251,20 @@ fixtures = {
     // Creates a client, and access and refresh tokens for user 3 (author)
     createTokensForUser: function createTokensForUser() {
         return db.knex('clients').insert(DataGenerator.forKnex.clients).then(function () {
-            return db.knex('accesstokens').insert(DataGenerator.forKnex.createToken({user_id: 3}));
+            return db.knex('accesstokens').insert(DataGenerator.forKnex.createToken({id:3, user_id: 3}));
```
Comment (acburdine) [inline core/test/utils/index.js]: wouldn't it make more sense to have the `createTokensForUser` ids be 1 and 2, and then the `createMoreTokens` ids be 3 and 4?

**Comment by ramiel** [comment]
Yes, it is. But I let the code like that for now to have one thing less to check. I'll change it when the test will pass.
I even found a couple of errors in `models/base/token` where the promises are bad returned but I'll open a new PR for this

File: core/test/utils/index.js (hunk 1 above)
Comment (ramiel) [inline core/test/utils/index.js]: Yes, it is. But I let the code like that for now to have one thing less to check. I'll change it when the test will pass.
I even found a couple of errors in models/base/token where the promises are bad returned but I'll open a new PR for this

**Comment by ramiel** [comment]
I think is finished now. The broken test on travis looks random to me, not related to this code but if you find it is related just let me know.

**Comment by sebgie** [comment]
I'm sorry, but this doesn't work for me at all :-(.

For testing I'm logged in Chrome and Safari with the same user. When I change the password in Chrome, go back to Safari, refresh I'm logged out as expected. If I then refresh Chrome I'm also logged out. From a quick look into the database I could see that all access tokens are destroyed when changing the password.

Looking deeper into this issue I think that this is more complex than originally anticipated and ember-simple-auth is also not behaving as expected.

- *ember-simple-auth*: When the access token becomes invalid I would expect the library to fetch a new access token using the refresh token, if that fails the user is logged out. Instead the user is logged out immediately after not being able to authenticate using the access token? Is this intended or a bug? cc: @acburdine @kevinansfield 

- *invalidate other logins*: If a refresh token is present it could be used to generate a new/valid access token. That means that we'll also need to invalidate all existing refresh tokens except the one that is associated with the access token used to change the password. We don't keep track of this information atm.

It seems to me like we need to look into the ember-simple-auth issue first and implement some database changes before it is viable to finish this feature?

**Comment by kevinansfield** [comment]
@sebgie I think ESA will only try to refresh the token if it's known to be expired. If the token is deleted on the server then it means ESA has an invalid token that as far as it knows should still be active and so it treats it as though the session has ended - I think that's correct as it prevents tampering and the ability to resurrect sessions that should be dead?

**Comment by ErisDS** [comment]
Closing this until #6626 is done.

**Pull request closed without merging**